File: backend/src/main/java/VoidSystems/appointment_service/service/AvailabilityService.java
package VoidSystems.appointment_service.service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.List;

import VoidSystems.appointment_service.dto.availability.AvailabilityDTO;

public interface AvailabilityService {
    
    List<AvailabilityDTO> getAllAvailabilities();
    
    List<AvailabilityDTO> getAvailabilitiesByProviderId(Long providerId);
    
    List<AvailabilityDTO> getRecurringAvailabilitiesByProviderId(Long providerId);
    
    List<AvailabilityDTO> getNonRecurringAvailabilitiesByProviderId(Long providerId);
    
    List<AvailabilityDTO> getAvailabilitiesByProviderIdAndDayOfWeek(Long providerId, DayOfWeek dayOfWeek);
    
    List<AvailabilityDTO> getAvailabilitiesByProviderIdAndDate(Long providerId, LocalDate date);
    
    AvailabilityDTO getAvailabilityById(Long id);
    
    AvailabilityDTO createAvailability(AvailabilityDTO availabilityDTO, Long providerId);
    
    AvailabilityDTO updateAvailability(Long id, AvailabilityDTO availabilityDTO);
    
    void deleteAvailability(Long id);
    
    boolean isTimeSlotAvailable(Long providerId, LocalDate date, java.time.LocalTime startTime, java.time.LocalTime endTime);
}

File: frontend/src/api/availabilityService.ts
import axios from 'axios';
import { API_BASE_URL } from '../config';

// Types
export interface Availability {
  id: number;
  providerId: number;
  providerName?: string;
  recurring: boolean;
  dayOfWeek?: number;
  specificDate?: string;
  startTime: string;
  endTime: string;
  createdAt?: string;
  updatedAt?: string;
}

// Create axios instance with base URL
const api = axios.create({
  baseURL: API_BASE_URL,
});

// Add request interceptor to include auth token
api.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Availability service methods
export const availabilityService = {
  // Get all availabilities (admin only)
  getAllAvailabilities: () => {
    return api.get<{ success: boolean; data: Availability[] }>('/availabilities');
  },
  
  // Get availability by ID
  getAvailabilityById: (availabilityId: number) => {
    return api.get<{ success: boolean; data: Availability }>(`/availabilities/${availabilityId}`);
  },
  
  // Get availabilities by provider ID
  getAvailabilitiesByProviderId: (providerId: number) => {
    return api.get<{ success: boolean; data: Availability[] }>(`/availabilities/provider/${providerId}`);
  },
  
  // Get recurring availabilities by provider ID
  getRecurringAvailabilitiesByProviderId: (providerId: number) => {
    return api.get<{ success: boolean; data: Availability[] }>(`/availabilities/provider/${providerId}/recurring`);
  },
  
  // Get specific date availabilities by provider ID
  getSpecificDateAvailabilitiesByProviderId: (providerId: number, date: string) => {
    return api.get<{ success: boolean; data: Availability[] }>(`/availabilities/provider/${providerId}/date/${date}`);
  },
  
  // Create availability (provider only)
  createAvailability: (providerId: number, availabilityData: Omit<Availability, 'id' | 'providerId' | 'createdAt' | 'updatedAt'>) => {
    return api.post<{ success: boolean; data: Availability }>(`/availabilities/provider/${providerId}`, availabilityData);
  },
  
  // Update availability
  updateAvailability: (availabilityId: number, availabilityData: Partial<Availability>) => {
    return api.put<{ success: boolean; data: Availability }>(`/availabilities/${availabilityId}`, availabilityData);
  },
  
  // Delete availability
  deleteAvailability: (availabilityId: number) => {
    return api.delete<{ success: boolean; message: string }>(`/availabilities/${availabilityId}`);
  },
};